### PR TITLE
Character limit label

### PR DESCRIPTION
**Pull request by BartoszKedziorek**
This PR refers to this issue: https://github.com/OpenLoco/OpenLoco/issues/1608

New label and counter, whichs allow user to see the characters limit and how many space is left.
**Changes**:
* new strings in language files
* new stringID
* new field and method in InputSession class
* changes in `OpenLoco::Ui::Windows::TextInput::draw`

**Example:**
picture1
picture2

I was thinking about putting it between bottom-left corner of textbox and "ok" button, but in my opinion it looks good as it is.

### CODE REVIEW

File: data/language/fr-FR.yml
Comment (duncanspumpkin) [inline data/language/fr-FR.yml:2186]: instead of hardcoding the character count just make it a parameter as text input windows have different limits.

File: src/OpenLoco/Windows/TextInputWindow.cpp
Comment (duncanspumpkin) [inline src/OpenLoco/Windows/TextInputWindow.cpp:246]: Do `Ui::Point(window.x + widget->left, window.y + widget->top - 10)` to take care of this

File: src/OpenLoco/Windows/TextInputWindow.cpp
Comment (duncanspumpkin) [inline src/OpenLoco/Windows/TextInputWindow.cpp:247]: This is pretty old code you have copied I would just do it as follows:
```cpp
auto args = FormatArguments();
args.push<uint16_t>(numCharacters);
args.push<uint16_t>(maxNumCharacters)
Gfx::drawStringLeft(*rt, &position, Colour::black, StringIds::num_characters_left_int_int, &args);
// Where StringIds::num_characters_left_int_int is "{COLOUR WINDOW_2}{UINT16}/{UINT16}"
```
This is much more flexible and doesn't involve messing about copying pointers of data and such.

File: data/language/en-GB.yml
Comment (duncanspumpkin) [inline data/language/en-GB.yml:2318]: You can get away with only updating en-GB all the other languages will fallback to using en-GB if there is no entry and as this is just displaying a number they will all be the same.

File: data/language/fr-FR.yml
Comment (BartoszKedziorek) [reply data/language/fr-FR.yml]: In the 3th commit I've changed it in the language files, but I'm pretty sure that limit in windows like: "changing town name", "scenario details", "scenario name", etc, is still hardcoded here:
https://github.com/OpenLoco/OpenLoco/blob/29d1deea7a3c7f5f652fc519bb4156aef29c1d02/src/OpenLoco/Ui/TextInput.cpp#L12-L19

Comment (duncanspumpkin) [reply data/language/fr-FR.yml]: I was thinking about the change company name window that has a 31 character limit. Not sure how that integrates into this as I'm pretty sure the text input let's you enter way more then it truncates later in the process. Have a look into it but we can leave it for future if it looks complex.

**Comment by duncanspumpkin** [comment]
Looks good. I've not tested it but can't see why it wouldn't work.

File: src/OpenLoco/Ui/TextInput.h
Comment (LeftofZen) [inline src/OpenLoco/Ui/TextInput.h:25]: where is the 199 coming from? it should be in a constant somewhere?

**Comment by LeftofZen** [comment]
this should get a changelog

**Review by LeftofZen — DISMISSED** [review]
- remove unused method, or use it
- use constant instead of hardcoded 199
- add changelog

File: data/language/fr-FR.yml
Comment (LeftofZen) [reply data/language/fr-FR.yml]: Agreed, the 199 is misleading since many other text entry fields don't actually accept 199 characters. @BartoszKedziorek any chance you can get the character limit to reflect the actual limit for that field, not the generic textbox entry limit? It'd probably involve a lot of work tbh, perhaps you could make a PR to enforce textbox length limits based on the field it's setting, and then this PR would be merged after it and you'd be able to get the correct lengths. Just an idea

**Comment by AaronVanGeffen** [comment]
I'd show the character count left of the 'OK' button. Leave the input box as-is, full-width.

**Comment by LeftofZen** [comment]
> I'd show the character count left of the 'OK' button, below the input box. Leave the input box as-is, full-width.
> 
> You could use the right-align text drawing function to this end as well.

I like that actually, looks nice
image

**Comment by duncanspumpkin** [comment]
> This begs the question if character limits are actually set correctly, as company names are actually limited to 32 characters.

31 characters as you need the null

File: data/language/fr-FR.yml
Comment (BartoszKedziorek) [reply data/language/fr-FR.yml]: > any chance you can get the character limit to reflect the actual limit for that field, not the generic textbox entry limit?
 It'd probably involve a lot of work tbh, perhaps you could make a PR to enforce textbox length limits based on the field it's setting, and then this PR would be merged after it and you'd be able to get the correct lengths.

Yeah, sure. I'm going to make your proposed PR first and after it is merged, I'll commit necessary changes to this PR.

File: src/OpenLoco/Ui/TextInput.cpp
Comment (AaronVanGeffen) [reply src/OpenLoco/Ui/TextInput.cpp:170]: Indeed, it looks like this method can be removed. The logic looks incorrect, either way, because the `cursorPosition` reflects the current cursor position (which can be moved), not the string length.

File: CHANGELOG.md
Comment (AaronVanGeffen) [inline CHANGELOG.md:6]: This warrants being marked a feature rather than a (bug)fix, I think. Could you move it to the top as well? (Features go above changes.)

Comment (BartoszKedziorek) [reply CHANGELOG.md:6]: Sorry, didn't mean to mark it as a fix. I'll change it in a second.

**Comment by AaronVanGeffen** [comment]
I've resolved the conflicts this PR had as a result of merging #1651, and moved the label to look as proposed:

image

**Review by AaronVanGeffen — APPROVED** [review]

**Comment by BartoszKedziorek** [comment]
@AaronVanGeffen @duncanspumpkin @LeftofZen. Thank you for all your reviews and help guys. Appreciate it :)

**Comment by LeftofZen** [comment]
@BartoszKedziorek my pleasure, thanks for a great feature and nice PR. Just a quick note for future PRs, after it's merged you can update the milestone to the next release as I did above ^

**Comment by AaronVanGeffen** [comment]
> after it's merged you can update the milestone to the next release as I did above

I believe only maintainers can do this, fwiw.

**Comment by LeftofZen** [comment]
> I believe only maintainers can do this, fwiw.

Oh I didn't realise, sorry! I'm happy to add the versions anyways

**Pull request merged**
